lib/board.rb

```diff
@@ -0,0 +1,57 @@
+# Tic Tac Toe Game Class
+class Board
+  def initialize
+    @board = Array.new(3).map { ['   '] * 3 }
+  end
+
+  def show_board
+    puts "\n      0     1     2     "
+    @board.each_index do |i|
+      puts '         |     |     '
+      print " #{i}   #{@board[i][0].nil? ? ' ' : @board[i][0]} |"
+      print " #{@board[i][1].nil? ? ' ' : @board[i][1]} |"
+      puts " #{@board[i][2].nil? ? ' ' : @board[i][2]}"
+      puts '    _____|_____|_____' unless i == 2
+    end
+    puts '         |     |     '
+    puts ''
+  end
+
+  def insert(field, turn)
+    first = field[0].to_i
+    second = field[1].to_i
+
+    if @board[first][second] == '   '
+      @board[first][second] = turn == 1 ? ' O ' : ' X '
+      return true
+    else
+      return false
+    end
+  end
+
+  def game_ends
+    count = 0
+    @board.each do |e|
+      count += 1 if e.include?('   ')
+    end
+
+    count.zero? || check_winner
+  end
+
+  def check_winner
+    combos = [
+      [@board[0][0], @board[0][1], @board[0][2]],
+      [@board[1][0], @board[1][1], @board[1][2]],
+      [@board[2][0], @board[2][1], @board[2][2]],
+      [@board[0][0], @board[1][0], @board[2][0]],
+      [@board[0][1], @board[1][1], @board[2][1]],
+      [@board[0][2], @board[1][2], @board[2][2]],
+      [@board[0][0], @board[1][1], @board[2][2]],
+      [@board[2][0], @board[1][1], @board[0][2]],
+    ]
```

Comment:
Can you just give us a hint on how to generate the combos dynamically? We are really struggling on this. Thanks in advance.

Reply:
This can go in multiple ways:
1. Generate all the index combinations:
- Something like:
```
Array.new(3).map.with_index do |row_elem, i|
  Array.new(3).map.with_index {|col_elem, j| [i,j]}
end
```

generates:

```
[[[0, 0], [0, 1], [0, 2]], 
[[1, 0], [1, 1], [1, 2]], 
[[2, 0], [2, 1], [2, 2]]]
```

You can play with the indexes `i,j` to generate the combinations you need. Modulo operator `%` is also really good for this kind of thing.

Reply:
Another option I see is to have multiple methods that map over the original matrix and return an array of arrays. Once you've mapped them you can just iterate over each array. 
E.g
```
columns = a.map.with_index do |row, row_idx|
  a.map {|row| row[row_idx]}
end
```

Once you have that you can do something lilke `columns.inject(true) {|memo, col| memo && col.uniq.length == 1}`

You can do this with rows, columns and also create a method to obtain the diagonals.

Reply:
Thank you for your hint, but we already now how to get the rows and the columns.
The problem is with creating the diagonals. We can't figure out a way how to dynamically produce them. And we've been trying and searching for so long. Could you actually show us?

Reply:
Left diagonal is kinda trivial. Think of the indexes. If you iterate over every row, you need elem 0 of row 0, elem 1 of row 1 and elem 2 of row 2. This should do the trick:
`a.map.with_index{|row, row_idx| p row[row_idx] }`

Right diagonal is a little bit trickier. The above solution shoud help 😉
Reach over the student dashboard if you get stuck.